src/nccl_ofi_net.c

```diff
@@ -474,7 +529,7 @@ static ncclResult_t register_mr_buffers(ofiComm_t *comm, void *data,
 	}
 
 	/* Check if provider requires registration of cuda device buffers */
-	if ((hmem_mr != true) && (type == NCCL_PTR_CUDA)) {
+	if ((hmem_mr != true) && (type == NCCL_PTR_CUDA) && (endpoint_mr != true)) {
```

Comment:
Why is the last condition required? It is possible to register CUDA buffers where memory is registered with the domain, right?

Reply:
I think this is a temporary workaround that I forgot to clean up. Given that we're setting `FI_REMOTE_READ` on `mr_attr.access` we need to register memory every time and ought to remove this check. Looking at the `fi_mr` manpage, it seems like we can only skip registration when the buffer is only accessed locally:
> If FI_MR_HMEM is set, the application must register buffers that were allocated using a device call and provide a valid desc parameter into applicable data transfer operations even if they are only used for local operations (e.g. send and receive data buffers).

Reply:
@rashikakheria IIUC, updating `register_mr_buffers` to include `FI_REMOTE_READ` on the access bits, renders the `hmem_mr` optimization invalid because the registration is not strictly local access. I had added the `endpoint_mr` check as a temporary workaround because the provider I'm working with uses `FI_MR_ENDPOINT`, does not set `FI_MR_HMEM`, and does require that the buffer be registered.

Based on this, I'll push a patch that deletes `hmem_mr` since that seems like the right thing to do. Would appreciate your feedback.

Reply:
A follow-up thought on `FI_REMOTE_READ`. I think we need not set this when `type == NCCL_PTR_HOST`, right?

Reply:
The plugin is only using send/recv APIs for data transfer therefore, we shouldn't be registering buffers with `FI_REMOTE_READ`. Is there a reason we need to add it? The only case where plugin calls RMA operations is `fi_read` for flush operations but that should only require us to set `FI_READ` attribute.

Can you help me understand why does the provider with `FI_MR_ENDPOINT` would not set `FI_MR_HMEM`?

Reply:
We always need to register buffers that can be accessed using one-sided operations, since the peer will need the MR key to access the buffer. `FI_MR_LOCAL` MR mode indicates that the provider wants the application to register local buffers, `FI_MR_HMEM` says we only need to do the registration when the buffer is in device memory (and pass in a descriptor). Since the NCCL flush operation accesses the CUDA buffer through `fi_read`, we always need to register these buffers and this short circuit was deleted from the PR in an earlier commit (still shows up here because the comment predates the update to the branch).

Reply:
As discussed on another thread, we should add both `FI_REMOTE_READ` and `FI_READ` for MR attributes when registering memory buffers. I am still trying to understand why we need this patch? https://github.com/aws/aws-ofi-nccl/pull/145/commits/c039bd46de755ab92e1b96e4f9db781c28d2dc0b